### PR TITLE
Add ability to pause reconciliation of Provider/Configuration/Function instances

### LINKED ISSUE #4619: "paused" annotation not being respected for configurations
### What happened?
It appears that the "crossplane.io/paused: "true"" annotation is not being respected for configurations. 

### How can we reproduce it?
To reproduce this issue, follow these steps: create a configuration, add the "paused" annotation, and manually update a composition. After a reconciliation cycle of the configuration, the composition reverts back to its previous state.

### What environment did it happen in?
Crossplane version: v1.13.1

### PR DESCRIPTION
**Pull request by pedjak**
### Description of your changes

The reconciliation can be paused by adding `crossplane.io/paused: "true"`
to the requested object. Removing the annotation resumes its reconciliation.

Fixes #4619 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit **and** E2E tests for my change.
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR, if necessary.
- [x] Opened a PR updating the [docs], if necessary.

[contribution process]: https://git.io/fj2m9
[docs]: https://docs.crossplane.io/contribute/contribute

### CODE REVIEW

File: internal/controller/pkg/manager/reconciler.go
Comment (turkenh) [inline internal/controller/pkg/manager/reconciler.go:300]: I think we don't need this. 

After noticing that this is different than [managed reconciler](https://github.com/crossplane/crossplane-runtime/blob/89e89660b5c6237f14be6563e7b1f104a9ac3589/pkg/reconciler/managed/reconciler.go#L758-L767), I found this discussion which explains why we don't need to clean the conditions up: https://github.com/crossplane/crossplane-runtime/pull/357#discussion_r984071756

File: internal/controller/pkg/revision/reconciler.go
Comment (turkenh) [inline internal/controller/pkg/revision/reconciler.go:474]: We should move this up before `meta.wasDeleted` check so that the pause will be effective for package revisins being deleted. This also aligns with how the feature was implemented in the managed reconciler and claim reconciler.

File: internal/controller/rbac/provider/binding/reconciler.go
Comment (turkenh) [inline internal/controller/rbac/provider/binding/reconciler.go:168]: Same here, please move above `meta.WasDeleted`

File: internal/controller/rbac/provider/roles/reconciler.go
Comment (turkenh) [inline internal/controller/rbac/provider/roles/reconciler.go:261]: Same here, please move above `meta.WasDeleted`

**Review by turkenh — COMMENTED** [review]
Thanks @pedjak!

Do you already have plans for documenting this? I think we need something similar to [this](https://docs.crossplane.io/latest/concepts/managed-resources/#paused) but for packages. Please also note, there is an ongoing work with package documentation and may be we can include [there](https://github.com/crossplane/docs/pull/566) (or at least build on top of that)

File: internal/controller/pkg/revision/reconciler.go
Comment (pedjak) [reply internal/controller/pkg/revision/reconciler.go:474]: Meaning that we cannot remove paused objects?

**Comment by pedjak** [comment]
> Do you already have plans for documenting [this](https://docs.crossplane.io/latest/concepts/managed-resources/#paused)? I think we need something similar to this but for packages. Please also note, https://github.com/crossplane/docs/pull/566 is an ongoing work with package documentation and may be we can include there (or at least build on top of that)

We should certainly document this after it gets merged. /cc @plumbis

File: internal/controller/pkg/revision/reconciler.go
Comment (turkenh) [reply internal/controller/pkg/revision/reconciler.go:474]: Yes, deleted objects would still need reconciliation to finalize their deletion. 
One can remove a paused object after removing the paused annotation.

File: internal/controller/pkg/manager/reconciler.go
Comment (turkenh) [reply internal/controller/pkg/manager/reconciler.go:300]: Ah ok, I missed that package types not having a `Synced` condition. I am wondering if introducing a `Synced` annotation could make sense, but something to consider later.

File: internal/controller/pkg/revision/reconciler.go
Comment (pedjak) [reply internal/controller/pkg/revision/reconciler.go:474]: Is that really what users expect to happen? In [the docs](https://docs.crossplane.io/latest/concepts/managed-resources/#paused) we do not say anything about being impossible to remove paused objects.

Imagine the following situation: a user attempts to remove a namespace containing a paused claim. This operation would hangs forever, right?

Comment (turkenh) [reply internal/controller/pkg/revision/reconciler.go:474]: > Is that really what users expect to happen?

As said, deletion is also part of reconciliation, the controller needs some operation / clean-up to finalize deletion. I would be surprised that part of reconciliation was not "paused". And that could be more desirable if I want to debug before the resource is deleted.

Another point is, one can go and find why their resource not deleted if that is the behavior they are looking for and remove the annotation to proceed. But if we do the opposite, there wouldn't be a way to revert for a user expecting the deletion to hang on with paused annotation, as the resource would already be gone.

> Imagine the following situation: a user attempts to remove a namespace containing a paused claim. This operation would hangs forever, right?

They can see why it hangs and remove the paused annotation to proceed. 
I don't think this is a feature that people would end up having paused resources around. It is more like a deliberate action where people would revert after they get what they are after.

Comment (phisco) [reply internal/controller/pkg/revision/reconciler.go:474]: I agree with @turkenh, pausing should pause everything, it would be surprising otherwise.

Comment (sttts) [reply internal/controller/pkg/revision/reconciler.go:474]: I tend to be on @pedjak's side here. Broken deletion seems to be unexpected and strange.

Comment (pedjak) [reply internal/controller/pkg/revision/reconciler.go:474]: Just tried the following:
* created namespace `example`
* created claim in it
* paused claim
* tried to remove namespace with `kubectl delete ns example`
* `kubectl` hangs with the message `namespace "example" deleted` but it does not return to the prompt

How a user can know that there is a claim in the namespace holding the delete?

Even if I do `kubectl describe` on that claim, there is no trace saying that deletion is blocked and can be resumed when pause annotation is removed.

Users do not need to know anything about how things are working under the hood, how many controllers are processing objects, in what order, etc. However, when users want to remove an object, we should either protect against that with a proper error message, or allow him to do that. If an object reconciliation is paused, we can still allow removal, cleaning up what is needed, etc.

Comment (turkenh) [reply internal/controller/pkg/revision/reconciler.go:474]: Let me link the discussion on slack: https://crossplane.slack.com/archives/CEF5N8X08/p1697618937115119?thread_ts=1697555426.091559&cid=CEF5N8X08

> Even if I do `kubectl describe` on that claim, there is no trace saying that deletion is blocked and can be resumed when pause annotation is removed.

This is something different. There should be. Shouldn't this output some events/logs: https://github.com/crossplane/crossplane/blob/af2496ebacd6227f19be3b1fb8af58100a4d8b2e/internal/controller/apiextensions/claim/reconciler.go#L371-L379

As I stated in the thread, it feels like we are trying to optimize user experience for a very specific scenario while ignoring others who would expect the reconciliation to be paused, including the deletion.

FWIW, it is also implemented in the same way in cluster API, which could be considered as an indication that this is what most people expect from such a functionality, see: [https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/b90810b70864664a00314f40a3f8a31d4215f1b5/controllers/awscluster_controller.go#L[…]4](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/b90810b70864664a00314f40a3f8a31d4215f1b5/controllers/awscluster_controller.go#L165-L194)

Comment (pedjak) [reply internal/controller/pkg/revision/reconciler.go]: > This is something different. There should be. Shouldn't this output some events/logs: 

The events are gone, because when you try to remove a namespace, k8s starts removing all of them in the given namespace, including events. Given that there is no problems to remove them (no finalizers, no blockers), the events are all gone.

> it feels like you are trying to optimize user experience for a very specific scenario while ignoring others who would expect the reconciliation to be paused, including the deletion.

Who are the others, i.e. what is the usecase we want to support?

We indicate that reconciliation is paused via condition of type `Synced` like this:

```yaml
reason: ReconcilePaused
status: "False"
type: Synced
```

when `False` it means that we are not syncing claim changes with the counterpart composite, i.e. we are not passing claim updates down the line. However, object deletion is not an update, at least not in CRUD/REST context, and therefore an user should be very surprised to discover that deletion of paused objects or higher order objects get stuck.

In order to [explore/learn](https://github.com/crossplane/crossplane/issues/4839) more about the use case mentioned by @turkenh I have updated the PR to match the behavior already available for claims/composites. Furthermore, I have added to the pause status condition a human readable message explain the consequences, and in the case of deletion we generate an additional k8s event and a log entry on info level.

File: apis/apiextensions/v1/conditions.go
Comment (turkenh) [inline apis/apiextensions/v1/conditions.go:94]: I would suggest just opening a PR there, but given the feature freeze deadline, can you at least open an issue to follow up?

File: apis/apiextensions/v1/conditions.go
Comment (turkenh) [inline apis/apiextensions/v1/conditions.go:98]:
```suggestion
	c.Message = "Reconciliation (including deletion) is paused via the pause annotation"
```

File: internal/controller/apiextensions/claim/reconciler.go
Comment (turkenh) [inline internal/controller/apiextensions/claim/reconciler.go:381]:
```suggestion
		r.record.Event(cm, event.Normal(reasonPaused, "Reconciliation (including deletion) is paused via the pause annotation"))
		cm.SetConditions(v1.ReconcilePaused())
```

It should be enough to update the message as above, no need for special handling for `meta.wasDeleted`.

After this message, it is clear that:

- Reconciliation is paused.
- This includes deletion.
- It is via pause annotation, so, you can remove it to resume. It is the same for both resuming deleted and non-deleted resources.

File: apis/apiextensions/v1/conditions.go
Comment (pedjak) [reply apis/apiextensions/v1/conditions.go:94]: Actually, I just realized that I do not need to modify crossplane-runtime: on condition there is `WithMessage()` method so the human readable message can be provided easily.

File: internal/controller/apiextensions/claim/reconciler.go
Comment (pedjak) [reply internal/controller/apiextensions/claim/reconciler.go]: Updated the PR with the suggested changes. I assumed I should apply it to all places in this PR. Please let me know if my understanding was correct.

> It should be enough to update the message as above, no need for special handling for meta.wasDeleted.

It should be enough, but this degrades the previously attempted user experience improvements. Namely:
* the logs are generated on DEBUG level, making `Reconciliation (including deletion) is paused via the pause annotation"` message invisible by default. The previous attempt wanted to bring them to INFO level, so that users can quickly spot them in the logs
* the current message is rather vague, refers to `pause annotation`. IMHO, we should be as much as explicit to users, so that they do not need to checks docs to understand what pause annotation is. Writing the exact annotation in the message is super helpful when one person pauses reconciliation, but another one performs some cluster maintenance/cleanup. In large orgs that is a common case. That person does not even need to know what Crossplane is, but the right message would unblock them quickly.
* Having an extra message that deletion is attempted on paused object would further help by debugging.

Comment (turkenh) [reply internal/controller/apiextensions/claim/reconciler.go]: Thanks!

> - the logs are generated on DEBUG level, making Reconciliation (including deletion) is paused via the pause annotation" message invisible by default. The previous attempt wanted to bring them to INFO level, so that users can quickly spot them in the logs

I agree that INFO would be better here 👍 

> - the current message is rather vague, refers to pause annotation. IMHO, we should be as much as explicit to users, so that they do not need to checks docs to understand what pause annotation is. Writing the exact annotation in the message is super helpful when one person pauses reconciliation, but another one performs some cluster maintenance/cleanup. In large orgs that is a common case. That person does not even need to know what Crossplane is, but the right message would unblock them quickly.

Users can just check the annotations on the resource and figure out which is the pause annotation without necessarily checking the documentation. However, being more explicit wouldn't hurt.

> - Having an extra message that deletion is attempted on paused object would further help by debugging.

Just like the above, I would agree with this in the context of being more explicit. However, I don't think this is worth introducing an additional branch (i.e. `if`) to the logic.

**Review by turkenh — APPROVED** [review]

File: internal/controller/pkg/manager/reconciler.go
Comment (pedjak) [reply internal/controller/pkg/manager/reconciler.go]: done.

**Pull request merged**
